mlrun/api/db/sqldb/db.py

```diff
@@ -296,9 +300,28 @@ def list_runs(
                 max_partitions,
             )
 
+        # Purposefully not using outer join to avoid returning runs without notifications
+        if with_notifications:
+            query = query.join(Notification, Run.id == Notification.run)
+
         runs = RunList()
         for run in query:
-            runs.append(run.struct)
+            run_struct = run.struct
+            if with_notifications:
+                for notification in run.notifications:
+                    (
+                        notification_spec,
+                        notification_status,
+                    ) = self._transform_notification_record_to_spec_and_status(
+                        notification
+                    )
+                    run_struct.setdefault("spec", {}).setdefault(
+                        "notifications", []
+                    ).append(notification_spec)
+                    run_struct.setdefault("status", {}).setdefault("notifications", {})[
+                        notification.name
+                    ] = notification_status
```

Comment:
I think it worth moving the `setdefault` before the start of the loop that way the code inside would be cleaner, and anyway we would like to have the `notifications` attribute inside the object even if there are no notifications for that specific run.

Reply:
First part - 👍 
Regarding no notifications - because of the join above (and the comment explaining it) no run should return without notifications when querying `with_notifications=True`